File: 20-actors/rasen/methods/datom_emit.py
#!/usr/bin/env python3
"""rasen 螺旋 — kotoba Datom-log emitter (canonical EAVT state, ADR-2605312345).

Projects the public-genetics graph into append-only kotoba Datoms [e a v tx op] — the
first-class canonical state (NOT a projection cache). Two strata:

  GROUND (durable, op :add) — one datom per (entity, attribute, value): the gene / variant /
    phenotype / population / pathway nodes and the :en/* 縁. This IS the Datom log.

  DERIVED (transient, :bond/is-transient true) — the edge-primary care / burden / pleiotropy
    integrals. Per N1/G2 these are computed on READ and are NOT stored as ground datoms; they
    are emitted in a clearly-flagged transient block so a reader can materialise them without
    mistaking them for persisted state.

Pure stdlib — runnable inside the rasen kotoba pywasm actor (componentize-py).
Usage:
    python3 datom_emit.py [seed.edn] [--out OUTDIR] [--tx N]
"""
from __future__ import annotations
import sys, pathlib
sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from analyze import load, analyze, read_edn  # noqa: F401,E402

# attributes promoted from each node/edge map into ground datoms (stable order = determinism)
NODE_ATTRS = [":genome/kind", ":genome/label", ":genome/sourcing", ":genome/links",
              ":gene/symbol", ":gene/cytoband", ":gene/ensembl", ":gene/taxon",
              ":variant/rsid", ":variant/category",
              ":phenotype/code", ":phenotype/inheritance",
              ":population/code", ":pathway/source", ":pathway/acc"]
EDGE_ATTRS = [":en/from", ":en/to", ":en/kind", ":en/grasping-load", ":en/clinsig", ":en/sourcing"]


def _fmt(v) -> str:
    if v is True:
        return "true"
    if v is False:
        return "false"
    if v is None:
        return "nil"
    if isinstance(v, str):
        return v if v.startswith(":") else '"' + v.replace('\\', '\\\\').replace('"', '\\"') + '"'
    if isinstance(v, float):
        return f"{v:g}"
    return str(v)


def emit(nodes: dict, edges: list, res: dict, tx: int = 1) -> str:
    L = []
    L.append(";; rasen 螺旋 — GENERATED kotoba Datom log (ADR-2606101000). DO NOT hand-edit.")
    L.append(";; Canonical EAVT state (ADR-2605312345). [e a v tx op].")
    L.append(";; GROUND op :add = durable. DERIVED :bond/is-transient = computed on read (N1/G2).")
    L.append(";; G1: PUBLIC reference genetics only — no individual genotypes; aggregate frequencies only.")
    L.append("[")

    # ── GROUND: node datoms
    for nid in nodes:  # insertion order (EDN read order) → deterministic
        n = nodes[nid]
        for a in NODE_ATTRS:
            if a in n and n[a] is not None:
                L.append(f"[{_fmt(nid)} {a} {_fmt(n[a])} {tx} :add]")

    # ── GROUND: edge datoms (edge entity id is content-stable: en.<from>.<kind>.<to>)
    for e in edges:
        eid = f"en.{e[':en/from']}.{e[':en/kind'].lstrip(':')}.{e[':en/to']}"
        for a in EDGE_ATTRS:
            if a in e and e[a] is not None:
                L.append(f"{'[' + _fmt(eid)} {a} {_fmt(e[a])} {tx} :add]")

    # ── DERIVED (transient — NOT persisted; N1/G2)
    L.append(";; ── DERIVED readouts (transient; integral of incident 縁, computed on read) ──")
    for nid, v in sorted(res["care"].items(), key=lambda kv: (-kv[1], kv[0])):
        L.append(f"[{_fmt(nid)} :bond/care-priority {v:g} {tx} :derived] ;; :bond/is-transient true")
    for nid, v in sorted(res["burden"].items(), key=lambda kv: (-kv[1], kv[0])):
        L.append(f"[{_fmt(nid)} :bond/locus-burden {v:g} {tx} :derived] ;; :bond/is-transient true")
    for nid, v in sorted(res["pleiotropy"].items(), key=lambda kv: (-kv[1], kv[0])):
        L.append(f"[{_fmt(nid)} :bond/pleiotropy {v:g} {tx} :derived] ;; :bond/is-transient true")

    L.append("]")
    return "\n".join(L) + "\n"


def main(argv):
    here = pathlib.Path(__file__).resolve().parent.parent
    seed = pathlib.Path(argv[1]) if len(argv) > 1 and not argv[1].startswith("--") \
        else here / "data" / "seed-genome-graph.kotoba.edn"
    outdir = here / "out"
    if "--out" in argv:
        outdir = pathlib.Path(argv[argv.index("--out") + 1])
    tx = int(argv[argv.index("--tx") + 1]) if "--tx" in argv else 1
    outdir.mkdir(parents=True, exist_ok=True)

    nodes, edges = load(seed)
    res = analyze(nodes, edges)
    out = outdir / "genome-datoms.kotoba.edn"
    out.write_text(emit(nodes, edges, res, tx), encoding="utf-8")
    print(f"rasen datom log → {out} ({len(nodes)} nodes + {len(edges)} 縁, tx={tx})")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: 20-actors/kuramori/methods/datom_emit.clj
;; kuramori 倉守 — kotoba Datom-log emitter (canonical EAVT state, ADR-2605312345).
;;
;; Projects the warehouse graph into append-only kotoba Datoms [e a v tx op].
;;   GROUND (op :add, durable) — zone / slot / sku / robot node datoms + placement 縁.
;;     This IS the Datom log.
;;   DERIVED (op :derived, transient :bond/is-transient true) — slotting weighted-travel,
;;     dispatch makespan, battery gate; computed on READ, NOT persisted (N1/G2 pattern,
;;     mirrors asobi/datom_emit).
;;
;; Pure Clojure, no deps → babashka-runnable AND kotoba-pywasm-portable.
;; Per ADR-2606142000 (kuramori R0).
(ns kuramori.methods.datom-emit
  (:require [clojure.edn :as edn]
            [clojure.string :as str]
            [kuramori.methods.analyze :as az]))

(defn fmt
  "Format a value as an EDN Datom field: keywords bare, strings quoted, bools/nil literal."
  [v]
  (cond
    (true? v) "true"
    (false? v) "false"
    (nil? v) "nil"
    (keyword? v) (str v)
    (string? v) (if (str/starts-with? v ":")
                  v
                  (str \" (str/escape v {\\ "\\\\" \" "\\\""}) \"))
    (float? v) (format "%g" v)
    :else (str v)))

(defn datom [e a v tx op]
  (str "[" (fmt e) " " (str a) " " (fmt v) " " tx " " (str op) "]"))

(defn emit
  "Emit the warehouse Datom log as an EDN string. `seed` is the loaded map,
   `res` the analyze/run result, `tx` the transaction number."
  [seed res tx]
  (let [L (atom [])
        add! (fn [s] (swap! L conj s))]
    (add! ";; kuramori 倉守 — GENERATED kotoba Datom log (ADR-2606142000). DO NOT hand-edit.")
    (add! ";; Canonical EAVT state (ADR-2605312345). [e a v tx op].")
    (add! ";; GROUND op :add = durable. DERIVED :bond/is-transient = computed on read (N1/G2).")
    (add! "[")
    ;; GROUND — zones
    (doseq [z (:zones seed)]
      (add! (datom (:id z) :wh.zone/kind (:kind z) tx :add)))
    ;; GROUND — slots (+ slot-in-zone 縁)
    (doseq [s (:slots seed)]
      (add! (datom (:id s) :wh.slot/dist-from-face (double (:dist-from-face s 0)) tx :add))
      (add! (datom (:id s) :wh.slot/max-kg (:max-kg s 0) tx :add))
      (add! (datom (:id s) :wh.slot/in-zone (:zone s) tx :add)))
    ;; GROUND — SKUs (+ ABC class)
    (doseq [k (:skus seed)]
      (add! (datom (:id k) :wh.sku/velocity (:velocity k 0) tx :add))
      (add! (datom (:id k) :wh.sku/weight-kg (:weight-kg k 0) tx :add))
      (add! (datom (:id k) :wh.sku/abc (get-in res [:abc (:id k)]) tx :add)))
    ;; GROUND — robots (fleet)
    (doseq [r (:fleet seed)]
      (add! (datom (:id r) :wh.robot/kind (:kind r) tx :add)))
    ;; GROUND — placement 縁 (sku → slot)
    (doseq [[sku-id slot-id] (get-in res [:slotting :placement])]
      (add! (datom (str "en." sku-id ".slotted-in." slot-id) :en/from sku-id tx :add))
      (add! (datom (str "en." sku-id ".slotted-in." slot-id) :en/to slot-id tx :add))
      (add! (datom (str "en." sku-id ".slotted-in." slot-id) :en/kind :slotted-in tx :add)))
    ;; DERIVED — transient readouts (computed on read; not durable)
    (add! ";; ── DERIVED readouts (transient; computed on read) ──")
    (add! (datom (:id (:facility seed)) :bond/weighted-travel
                 (double (get-in res [:slotting :weighted-travel])) tx :derived))
    (add! (datom (:id (:facility seed)) :bond/dispatch-makespan
                 (double (get-in res [:dispatch :makespan])) tx :derived))
    (add! (datom (:id (:facility seed)) :bond/charge-needed
                 (get-in res [:battery :charge-needed]) tx :derived))
    (add! "]")
    (str (str/join "\n" @L) "\n")))

(defn -main [& args]
  (let [path (or (first args) "20-actors/kuramori/data/warehouse.edn")
        seed (az/load-seed path)
        res (az/run seed)]
    (print (emit seed res 1))
    (flush)))
